book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,497 @@
+
+```rust
+pub struct PreparedBlock {
+    ..
+    /// The value balances for each pool for this block.
+    pub block_value_balance: ValuePool<NegativeAllowed>,
+}
+```
+- In `zebra-consensus/src/block.rs` pass the value balance to the zebra-state:
+
+```rust
+let block_value_balance = block.value_balance();
+let prepared_block = zs::PreparedBlock {
+    ..
+    block_value_balance,
+};
+```
+
+### Add a value pool into the state `Chain` struct
+
+- This is the value pool for the non finalized part of the blockchain.
+- Location of the `Chain` structure where the pool field will be added: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+pub struct Chain {
+    ..
+    /// The chain value pool balance at the tip of this chain.
+    value_pool: ValueBalance<NonNegative>,
+}
+```
+- Add a new argument `finalized_tip_value_balance` to the `commit_new_chain()` method located in the same file.
+- Pass the new argument to the Chain in:
+
+```rust
+let mut chain = Chain::new(finalized_tip_history_tree, finalized_tip_value_balance);
+```
+
+Note: We don't need to pass the finalized tip value balance into the `commit_block()` method.
+
+### Check the consensus rules when the chain is updated or reversed
+
+- Location: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+impl UpdateWith<PreparedBlock> for ValueBalance<NegativeAllowed> {
+    fn update_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
+        self += value_balance?;
+    }
+
+    fn revert_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
+        self -= value_balance?;
+    }
+}
+```
+
+### Changes to finalized state
+
+The state service will call `commit_new_chain()`. We need to pass the value pool from the disk into this function.
+
+```rust
+self.mem
+    .commit_new_chain(prepared, self.disk.history_tree(), self.disk.get_pool())?;
+```
+
+We now detail what is needed in order to have the `get_pool()` method available.
+
+#### Serialization of `ValueBalance<C>`
+
+In order to save `ValueBalance` into the disk database we must implement `IntoDisk` and `FromDisk` for `ValueBalance` and for `Amount`:
+
+```rust
+impl IntoDisk for ValueBalance<C> {
+    type Bytes = [u8; 32];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        [self.transparent.to_bytes(), self.sprout.to_bytes(),
+        self.sapling.to_bytes(), self.orchard.to_bytes()].concat()
+    }
+}
+
+impl FromDisk for ValueBalance<C> {
+    fn from_bytes(bytes: impl AsRef<[u8]>) -> Self {
+        let array = bytes.as_ref().try_into().unwrap();
+        ValueBalance {
+            transparent: i64::from_be_bytes(array[0..8]).try_into().unwrap()
+            sprout: i64::from_be_bytes(array[8..16]).try_into().unwrap()
+            sapling: i64::from_be_bytes(array[16..24]).try_into().unwrap()
+            orchard: i64::from_be_bytes(array[24..32]).try_into().unwrap()
```

Comment:
Bug: `to_bytes` is little endian, but `from_bytes` is big endian.

Let's put the deserialization in `Amount::from_bytes`, so the serialization and deserialization code is in the same file? If we do that, bugs like this will be more obvious.

Can you also add a proptest for `IntoDisk`/`FromDisk` for `ValueBalance` that would catch this bug?

Reply:
I fixed this at https://github.com/ZcashFoundation/zebra/pull/2430/commits/8756c9200fc727805a4c3e5c8906313d13851871